Veli JS3-wk1

veliataseven opened the pull request.

remarcmij commented on homework/index.js:
```diff
@@ -30,13 +30,162 @@
+
+    const defaultRepo = defaultSelectedRepository(data, select[select.selectedIndex].text);
+
+    function showSelectedConts(dataCont) {
```
You should avoid nested functions like this. The reason is that nested function obscure where the data comes from that they need. A nested function has access to the data of its outer function. In this case, the `showSelectedConst` function accesses the `container` variable from its outer function, but that you would have to look hard to understand that. If you place this function outside of the `buildHtml` function you immediately see the `container` variable being red underlined in VSCode, indicating that it is now not defined. To solve that, you can pass it through the parameter list, making it absolutely clear what this function needs from the outside world to do its work:

```js
function showSelectedConts(dataCont, container) {
 ...
}

function buildHtml(data, root) {
  ...
  showSelectConts(contributors, container);
  ...
}
```

BTW, avoid unnecessary abbreviations like `Conts` in `showSelectedConts`. It is only a few more characters to make it `showSelectedContributors` and so much nicer to read without needing to guess.

remarcmij commented on homework/index.js:
```diff
@@ -30,13 +30,162 @@
+      }
+    });
+
+    function showSelectedRepo(selectedRepo) {
```
Place this function outside of the `buildHtml` function and pass it the `tbody` value that it needs through the parameter list:

```js
function showSelectedRepo(selectedRepo, tbody) {
  ...
}

function buildHtml(data, root) {
  ...
  showSelectedRepo(defaultRepo, tbody);
  ...
}

```

remarcmij commented on homework/index.js:
```diff
@@ -30,13 +30,162 @@
+        const tr4 = createAndAppend('tr', tbody);
+        createAndAppend('td', tr4, { class: 'label', text: 'Updated :' });
+        const updated = createAndAppend('td', tr4, { text: '' });
+        updated.textContent = new Date(selectedRepo.updated_at).toLocaleString();
```
1. Nice use of `toLocaleString()` to format the date according to the user's local computer settings. 
2. You can set the `textContent` directly by using it as the value of the `text` property in the previous line.

remarcmij commented on homework/index.js:
```diff
@@ -30,13 +30,162 @@
+        createAndAppend('td', tr1, { class: 'label', text: 'Repository :' });
+        const td2 = createAndAppend('td', tr1);
+        const repoLink = createAndAppend('a', td2, { href: '', target: '_blank', text: '' });
+        repoLink.textContent = selectedRepo.name;
```
You can set the `textContent` and `href` directly by using the `options` argument from the `createAndAppend` function (it was designed specifically for that purpose):

```js
const repoLink = createAndAppend('a', td2, {
  href: selectedRepo.html_url,
  target: '_blank',
  text: selectedRepo.name,
});
```

As an added bonus, if you do that you can also no longer need the `repoLink` variable: one less variable to worry about:

```js
createAndAppend('a', td2, {
  href: selectedRepo.html_url,
  target: '_blank',
  text: selectedRepo.name,
});
``` 

You can do similar things in other parts of your code.

veliataseven commented:
> Hi @veliataseven, you have working code and I can see you worked hard. But there are still a couple of fundamental improvements I would like you to make. Please check my comments below. The styling part is not urgent, you can do that in week 2 or 3.

thank you for detailed feedback..Actually I had already noticed that my codes has some problems that you mentioned . But Main problem is my bad studying timing.. I am gonna improve my codes according to your comments..I am gonna spare more time until now..

veliataseven commented:
> Hi @veliataseven, you have working code and I can see you worked hard. But there are still a couple of fundamental improvements I would like you to make. Please check my comments below. The styling part is not urgent, you can do that in week 2 or 3.

Hi Jim, I have done your comments.. => https://github.com/HackYourFuture/JavaScript3/pull/250/commits/594582dab1c0e47042a95146ad4eb78d31dfb766

remarcmij commented on homework/index.js:
```diff
@@ -30,18 +30,157 @@
+    repositories
+      .map(repository => repository.name)
+      .sort((a, b) => a.localeCompare(b, { caseFirst: 'lower' }))
+      .map((element, index) => createAndAppend('option', select, { value: index, text: element }));
```
I overlooked this in my first review: instead of sorting just an array of repository names (so that you can show those names in sorted order in the `<select>` element it is far more efficient to sort the complete array of repository objects. In that way there will be one-on-one correspondence between the index of the selected option from the `<select>` element and the index of the corresponding repository object:

```js
repositories.sort((a, b) => a.name.localeCompare(b.name));

repositories.forEach((repository, index) =>
  createAndAppend('option', select, { value: index, text: repository.name }),
);

const defaultRepository = repositories[0];
```

You now no longer need your function `getSelectedRepository` and can delete it.

remarcmij commented on homework/index.js:
```diff
@@ -30,18 +30,157 @@
+      const newSelectedRepository = getSelectedRepository(
+        repositories,
+        select[select.selectedIndex].text,
+      );
```
With reference to my comment below, you can replace lines 127-130 with:

```js
const newSelectedRepository = repositories[select.value];
```

veliataseven commented:
> Hi @veliataseven, thank you for addressing the feedback. There are a couple more comments below that I would like you to tackle along with the homework for week 2. For now I will approve this PR.

Thank you for great feedback @remarcmij  I have done your last comments. I am learning from you How to write clean and efficient code.. Last link is that https://github.com/HackYourFuture/JavaScript3/pull/250/commits/d8ad69062b8685f5924a17221e988cfb0a23cd2f
